### DISCUSSION TITLE
RGB vs GRB Listeners

### DISCUSSION THREAD

**Discussion by wrldtvlr** [josephdadams/TallyArbiter#320] [Q&A] [answered]
Turns out that certain listener devices use an alternate color order. Instead of RGB, they use GRB. Example - the M5 Matrix has a WS2812C driver, which expects GRB. There may be others.

Since the TA Server is sending the Pgm and Pvw colors to the client in RGB order, red and green were reversed when displayed on an M5 Matrix. I took a look at the server code, but didn't see any obvious/easy way to change the color order based on listener type.

This is easy enough to correct in the client, but I did have one question. I don't have an M5StickC, which looks like it might need RBG order and uses the same M5 client as the Matrix. Could someone that has a M5StickC confirm this? It may be necessary to create a separate client, or add logic to the existing client.

**Comment by hrueger** [comment]
I think this should be handled on client side. Unfortunately, I don't have an M5 either.

**Reply by MatteoGheza** [reply]
M5Atom/M5Stick needs RGB, so we convert the color client-side.
Here there are some examples of color conversion:

- M5AtomMatrix: https://github.com/josephdadams/TallyArbiter/blob/master/listener_clients/M5AtomMatrix-listener/tallyarbiter-m5atom/tallyarbiter-m5atom.ino#L320-L335
```cpp
actualColor.replace("#", "");
String hexstring = actualColor;
long colorNumber = (long) strtol( &hexstring[1], NULL, 16);
int g = colorNumber >> 16;
int r = colorNumber >> 8 & 0xFF;
int b = colorNumber & 0xFF;

if (actualType != "") {
  int backgroundColor = 0x10000 * r + 0x100 * g + b;
  int currColor[] = {backgroundColor, numbercolor};
  logger("Current color: " + String(backgroundColor), "info");
  //logger("Current camNumber: " + String(camNumber), "info");
  drawNumber(number[camNumber], currColor);
} else {
  drawNumber(number[camNumber], offcolor);
}
```  

- TTGO_T: https://github.com/josephdadams/TallyArbiter/blob/master/listener_clients/TTGO_T-listener/TTGO_T-listener.ino#L539-L551
```cpp
actualColor.replace("#", "");
String hexstring = actualColor;
long number = (long) strtol( &hexstring[1], NULL, 16);
int r = number >> 16;
int g = number >> 8 & 0xFF;
int b = number & 0xFF;
if (actualType != "") {
  tft.setTextColor(TFT_BLACK);
  tft.fillScreen(tft.color565(r, g, b));
} else {
  tft.setTextColor(TFT_WHITE);
  tft.fillScreen(TFT_BLACK);
}
```  

- Blink1: https://github.com/josephdadams/TallyArbiter/blob/master/listener_clients/blink1-listener/blink1-listener.py#L278
```python
current_color = hex_to_rgb(busses_list[0]["color"])
```

**Reply by wrldtvlr** [reply]
Yes, that's the change I made on the Matrix. I guess we'll see if anyone with an M5stickC has a problem.

There's a logger line that we'll fix as well.

**Reply by wrldtvlr** [reply]
@MatteoGheza  Actually, my reply makes no sense. Would you suggest we have separate clients for the M5StickC and M5 Matrix or add logic to the client to determine which order to use? Having one client makes the most sense to me if there's an easy way in code to identify the platform it's running on.

**Comment by wrldtvlr** [comment]
Ha! Looking through my workbench project box I discovered I have an M5StickC Plus. I'll see if I can play around with it and the Matrix and see what I can come up with.

**Comment by wrldtvlr** [comment]
I'm going to go lie down in the dark room and wait for the nice nurse to bring me my meds. I didn't realize there already were separate listener clients for the Matrix and M5StickC.

It's never easy though. My M5StickC has been sitting on the shelf for over a year now and apparently the internal battery has discharged, thus causing an Error 22 on startup. After some research, I discovered that a 300ohm resistor between the v3.3 and Batt pins will temporarily solve that problem. Hoping the battery will recharge in the not too distance future.

Next problem is a reference in the code to a library called Free_Fonts.h. There's no indication where this library is to be found.

Sigh.

**Reply by wrldtvlr** [reply]
Hmm. I thought I replied/followed up to this, but I don't see my post. I'm not sure it's the correct solution, but I found a copy of Free_Fonts.h on the web and saved it to my ..\Documents\Arduino\libraries\M5StickCPlus\src folder. That seems to work.

My Arduino development skills are limited. I'm sure there's a better way to identify what non-standard libraries are needed for any given sketch.

**Comment by wrldtvlr** [comment, marked as answer]
And I think we can mark this thread answered!
